### PR TITLE
U053, U054: Following and Followers 

### PR DESCRIPTION
**Pull request by enlne**
-does not update a user if they have been followed, but updates the user if they follow someone else

### CODE REVIEW

File: public/script.js [hunk 1]
```diff
@@ -76,8 +76,8 @@ function renderAccount(props) {
         spiritPointsCount: res.data.spiritPointsCount,
         magicPointsCount: res.data.magicPointsCount,
       });
-      allPlayers.setState({
-        showAllPlayers: false,
+      getPlayers.setState({
+        showGetPlayers: false,
```
Comment (Louise15926) [inline public/script.js]: variable naming: a name that starts with "show" sounds more like a function than a boolean variable. Something like "shouldShow..." might be better for a boolean.

File: e2e/following.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    await registerName.sendKeys('testName1');
+    await registerPassword.sendKeys('testPassword1');
+    await registerName.submit();
+
+    await driver.wait(until.elementLocated(By.css('.alert')));
+    const alert = await driver.findElement(By.css('.alert'));
+    assert.match(await alert.getText(), /Registered!/);
+  });
+
+  it('register2', async () => {
```
Comment (Louise15926) [inline e2e/following.js]: Seems like the two callbacks are reusable

File: app/src/PlayersList.js [hunk 3]
```diff
@@ -24,17 +25,25 @@ export default class PlayersList extends React.Component {
     }
   }
 
+  async followPlayer(playerTo) {
+    const followingList = await this.app.genericApi1('cl_following', playerTo);
```
Comment (Louise15926) [inline app/src/PlayersList.js]: Would probably be nice to have some error handling here. I see that genericApi1 is returning a promise, but i don't see you supplying a resolve/reject callback to the promise. (I understand that there might be reasons why you find it unnecessary to handle that; my comment is just for general cases)

File: public/script.js (hunk 1 above)
Comment (enlne) [reply public/script.js]: I agree. Will change the variable name.

**Comment by ra1nty** [comment]
Hi Dears,

Wanxian and I enjoyed reviewing your code and here is our checklist:

#### Coding Style
* Any redundant code? - Very few
* Does naming make sense - Makes sense
* Indentation & formatting - Overall good
* modularization - Some minor places need refactoring
#### Documentation
* Is code adequately documented - In my opinion should have some more comments.
#### Testing
* Does it test exception cases? - No
* Is coverage above 80%? - Yes
#### Handling
* Are exceptions properly handled? - Overall ok, we found some promises miss rejection handler and some exceptions unhandled

File: src/index.js [hunk 4]
```diff
@@ -268,6 +274,40 @@ io.on('connection', function(socket) {
     done(success(users));
   });
 
+  socket.on('cl_following', async (playerTo, done) => {
+    debug('cl_following', playerTo);
+
+    if (!user) return done(error('forbidden'));
```
Comment (Louise15926) [inline src/index.js]: If user = 0, !user would return true (which might be undesirable). In JavaScript, it is normally considered a good coding style to have if (user == null) to check whether a variable is null or undefined. Again, it is probably that you have maintained some kind of consistency throughout the other codes (e.g. user will never be an integer), and that having (!user) is effectively the same as having (user == null). My comment is just for general cases only for your reference.

Comment (enlne) [reply src/index.js]: Thank you for feedback. Indeed, user will never = 0.

**Comment by ra1nty** [comment]
Also in my opinion, since you already have React in your project stack, why not refactor those front-end html stuff, jQuery selectors into React components, that will improve the readability and consistency of the codebase.

**Comment by enlne** [comment]
> Also in my opinion, since you already have React in your project stack, why not refactor those front-end html stuff, jQuery selectors into React components, that will improve the readability and consistency of the codebase.

We very recently added React to our project so we are still in the process of converting our code.

File: e2e/following.js (hunk 2 above)
Comment (BINGXUHAN) [reply e2e/following.js]: We've refactored the register part

File: app/src/PlayersList.js (hunk 3 above)
Comment (Kailang) [reply app/src/PlayersList.js]: `this.app.genericApi1` has a default error handler defined in `App` class.
So, we do not need to handle the error again here.

**Pull request merged**